### PR TITLE
[Feature] Add typehint to mmdet3d/models/backbones.

### PR DESCRIPTION
**Pull request by Lum1104**
## Motivation

Add typehint to the code and make it more readable. 

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/models/backbones/base_pointnet.py [hunk 1]
```diff
@@ -1,25 +1,32 @@
...
+import torch
 from mmengine.model import BaseModule
 
 
 class BasePointNet(BaseModule, metaclass=ABCMeta):
     """Base class for PointNet."""
 
-    def __init__(self, init_cfg=None, pretrained=None):
+    def __init__(self,
+                 init_cfg: Optional[dict] = None,
```
Comment (sunjiahao1999) [inline mmdet3d/models/backbones/base_pointnet.py:14]: init_cfg use `OptMultiConfig` or `OptConfig` in `mmdet3d.utils`

File: mmdet3d/models/backbones/base_pointnet.py [hunk 2]
```diff
@@ -30,10 +37,11 @@ def _split_point_feats(points):
             torch.Tensor: Coordinates of input points.
             torch.Tensor: Features of input points.
         """
-        xyz = points[..., 0:3].contiguous()
+        xyz: torch.Tensor = points[..., 0:3].contiguous()
```
Comment (sunjiahao1999) [inline mmdet3d/models/backbones/base_pointnet.py:40]: No need in function implementation.

File: mmdet3d/models/backbones/dla.py [hunk 3]
```diff
@@ -9,7 +10,8 @@
 from mmdet3d.registry import MODELS
 
 
-def dla_build_norm_layer(cfg, num_features):
+def dla_build_norm_layer(cfg: Dict[str, Any],
```
Comment (sunjiahao1999) [inline mmdet3d/models/backbones/dla.py:13]: cfg use `OptMultiConfig` or `OptConfig` in `mmdet3d.utils`

File: mmdet3d/models/backbones/dla.py [hunk 4]
```diff
@@ -117,13 +121,13 @@ class Root(BaseModule):
...
-                 in_channels,
-                 out_channels,
-                 norm_cfg,
-                 conv_cfg,
-                 kernel_size,
-                 add_identity,
-                 init_cfg=None):
+                 in_channels: int,
+                 out_channels: int,
+                 norm_cfg: Dict[str, Any],
```
Comment (sunjiahao1999) [inline mmdet3d/models/backbones/dla.py:126]: same problem

File: mmdet3d/models/backbones/dla.py [hunk 5]
```diff
@@ -436,7 +443,7 @@ def _freeze_stages(self):
             for param in m.parameters():
                 param.requires_grad = False
 
-    def forward(self, x):
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, ...]:
```
Comment (sunjiahao1999) [inline mmdet3d/models/backbones/dla.py:446]: use `Tensor` rather than `torch.Tensor`.

File: mmdet3d/models/backbones/base_pointnet.py (hunk 1 above)
Comment (Lum1104) [reply mmdet3d/models/backbones/base_pointnet.py:14]: Thank you for your suggestion, I will revise it right away.

File: mmdet3d/models/backbones/dgcnn.py [hunk 6]
```diff
@@ -30,14 +33,16 @@ class DGCNNBackbone(BaseModule):
...
-                 radius=(None, None, None),
-                 gf_channels=((64, 64), (64, 64), (64, )),
-                 fa_channels=(1024, ),
-                 act_cfg=dict(type='ReLU'),
-                 init_cfg=None):
+                 in_channels: int,
+                 num_samples: Tuple[int, int, int] = (20, 20, 20),
+                 knn_modes: Tuple[str, str, str] = ('D-KNN', 'F-KNN', 'F-KNN'),
+                 radius: Tuple[Optional[float], Optional[float],
+                               Optional[float]] = (None, None, None),
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/dgcnn.py:40]:
```suggestion
                 radius: Sequence[Union[float, None]] = (None, None, None),
```

File: mmdet3d/models/backbones/dgcnn.py [hunk 7]
```diff
@@ -30,14 +33,16 @@ class DGCNNBackbone(BaseModule):
...
+                 gf_channels: Tuple[Tuple[int, int], Tuple[int, int],
+                                    Tuple[int]] = ((64, 64), (64, 64), (64, )),
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/dgcnn.py:42]:
```suggestion
                 gf_channels: Sequence[Sequence[int]] = ((64, 64), (64, 64), (64, )),
```

File: mmdet3d/models/backbones/dgcnn.py [hunk 8]
```diff
@@ -30,14 +33,16 @@ class DGCNNBackbone(BaseModule):
...
+                 fa_channels: Tuple[int] = (1024, ),
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/dgcnn.py:43]:
```suggestion
                 fa_channels: Sequence[int] = (1024, ),
```

File: mmdet3d/models/backbones/dgcnn.py [hunk 9]
```diff
@@ -30,14 +33,16 @@ class DGCNNBackbone(BaseModule):
...
+                 act_cfg: Dict = dict(type='ReLU'),
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/dgcnn.py:44]:
```suggestion
                 act_cfg: ConfigType = dict(type='ReLU'),
```

File: mmdet3d/models/backbones/dgcnn.py [hunk 10]
```diff
@@ -30,14 +33,16 @@ class DGCNNBackbone(BaseModule):
...
+                 init_cfg: Optional[Dict] = None):
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/dgcnn.py:45]:
```suggestion
                 init_cfg: OptMultiConfig = None):
```

File: mmdet3d/models/backbones/dgcnn.py [hunk 11]
```diff
@@ -47,10 +52,10 @@ def __init__(self,
 
         self.GF_modules = nn.ModuleList()
         gf_in_channel = in_channels * 2
-        skip_channel_list = [gf_in_channel]  # input channel list
+        skip_channel_list: List[int] = [gf_in_channel]  # input channel list
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/dgcnn.py:55]: This seems no need to add typehint. And please also check other places.

File: mmdet3d/models/backbones/dgcnn.py [hunk 12]
```diff
@@ -65,13 +70,15 @@ def __init__(self,
...
+        cur_fa_mlps: List[int] = list(fa_channels)
         cur_fa_mlps = [fa_in_channel] + cur_fa_mlps
 
         self.FA_module = DGCNNFAModule(
             mlp_channels=cur_fa_mlps, act_cfg=act_cfg)
 
-    def forward(self, points):
+    def forward(
+        self, points: torch.Tensor
+    ) -> Dict[str, Union[List[torch.Tensor], torch.Tensor]]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/dgcnn.py:81]:
```suggestion
    ) -> dict:
```

File: mmdet3d/models/backbones/dla.py [hunk 13]
```diff
@@ -1,15 +1,18 @@
...
 from mmengine.model import BaseModule
-from torch import nn
+from torch import Tensor, nn
 
 from mmdet3d.registry import MODELS
+from mmdet3d.utils import OptMultiConfig
 
 
-def dla_build_norm_layer(cfg, num_features):
+def dla_build_norm_layer(cfg: OptMultiConfig,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/dla.py:14]: OptMultiConfig -> ConfigType

File: mmdet3d/models/backbones/dla.py [hunk 14]
```diff
@@ -1,15 +1,18 @@
...
+                         num_features: int) -> Callable[[Any, int], Any]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/dla.py:15]:
```suggestion
                         num_features: int) -> Tuple[str, nn.Module]:
```

File: mmdet3d/models/backbones/dla.py [hunk 15]
```diff
@@ -53,13 +56,13 @@ class BasicBlock(BaseModule):
...
-                 in_channels,
-                 out_channels,
-                 norm_cfg,
-                 conv_cfg,
-                 stride=1,
-                 dilation=1,
-                 init_cfg=None):
+                 in_channels: int,
+                 out_channels: int,
+                 norm_cfg: Dict[str, Any],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/dla.py:61]:
```suggestion
                 norm_cfg: ConfigType,
```

**Review by Xiangxu-0103 — CHANGES_REQUESTED** [review]
Please use more standardized and appropriate typehints, you can take other modules as references.

File: mmdet3d/models/backbones/dgcnn.py (hunk 11 above)
Comment (Lum1104) [reply mmdet3d/models/backbones/dgcnn.py:55]: Thank you for your suggestion, I will fix it right now~

File: mmdet3d/models/backbones/multi_backbone.py [hunk 16]
```diff
@@ -27,16 +28,16 @@ class MultiBackbone(BaseModule):
...
-                 conv_cfg=dict(type='Conv1d'),
-                 norm_cfg=dict(type='BN1d', eps=1e-5, momentum=0.01),
-                 act_cfg=dict(type='ReLU'),
-                 suffixes=('net0', 'net1'),
-                 init_cfg=None,
-                 pretrained=None,
-                 **kwargs):
+                 num_streams: int,
+                 backbones: Union[List, Dict],
+                 aggregation_mlp_channels: Optional[List[int]] = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/multi_backbone.py:33]: Optional[Sequence[int]]

File: mmdet3d/models/backbones/multi_backbone.py [hunk 17]
```diff
@@ -27,16 +28,16 @@ class MultiBackbone(BaseModule):
...
+                 conv_cfg: Dict = dict(type='Conv1d'),
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/multi_backbone.py:34]: ConfigType. The same below.

File: mmdet3d/models/backbones/multi_backbone.py [hunk 18]
```diff
@@ -89,7 +90,7 @@ def __init__(self,
                           'please use "init_cfg" instead')
             self.init_cfg = dict(type='Pretrained', checkpoint=pretrained)
 
-    def forward(self, points):
+    def forward(self, points: torch.Tensor) -> Dict[str, List[torch.Tensor]]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/multi_backbone.py:93]:
```suggestion
    def forward(self, points: Tensor) -> dict:
```

File: mmdet3d/models/backbones/nostem_regnet.py [hunk 19]
```diff
@@ -59,15 +63,19 @@ class NoStemRegNet(RegNet):
         (1, 1008, 1, 1)
     """
 
-    def __init__(self, arch, init_cfg=None, **kwargs):
+    def __init__(self,
+                 arch: Dict[str, Union[int, float]],
+                 init_cfg: Optional[Dict] = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/nostem_regnet.py:68]: OptMultiConfig

File: mmdet3d/models/backbones/nostem_regnet.py [hunk 20]
```diff
@@ -59,15 +63,19 @@ class NoStemRegNet(RegNet):
...
 
-    def _make_stem_layer(self, in_channels, base_channels):
+    def _make_stem_layer(self, in_channels: int,
+                         base_channels: int) -> nn.Module:
         """Override the original function that do not initialize a stem layer
         since 3D detector's voxel encoder works like a stem layer."""
         return
 
-    def forward(self, x):
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, ...]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/nostem_regnet.py:78]: torch.Tensor -> Tensor

File: mmdet3d/models/backbones/pointnet2_sa_ssg.py [hunk 21]
```diff
@@ -31,20 +33,22 @@ class PointNet2SASSG(BasePointNet):
...
-                 num_points=(2048, 1024, 512, 256),
-                 radius=(0.2, 0.4, 0.8, 1.2),
-                 num_samples=(64, 32, 16, 16),
-                 sa_channels=((64, 64, 128), (128, 128, 256), (128, 128, 256),
-                              (128, 128, 256)),
-                 fp_channels=((256, 256), (256, 256)),
-                 norm_cfg=dict(type='BN2d'),
-                 sa_cfg=dict(
+                 in_channels: int,
+                 num_points: Tuple[int] = (2048, 1024, 512, 256),
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/pointnet2_sa_ssg.py:37]: Tuple ->Sequence

File: mmdet3d/models/backbones/pointnet2_sa_ssg.py [hunk 22]
```diff
@@ -31,20 +33,22 @@ class PointNet2SASSG(BasePointNet):
...
+                                                                     256)),
+                 fp_channels: Tuple[Tuple[int]] = ((256, 256), (256, 256)),
+                 norm_cfg: dict = dict(type='BN2d'),
+                 sa_cfg: dict = dict(
                      type='PointSAModule',
                      pool_mod='max',
                      use_xyz=True,
                      normalize_xyz=True),
-                 init_cfg=None):
+                 init_cfg: Optional[Dict[str, Any]] = None):
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/pointnet2_sa_ssg.py:51]: OptMultiConfig

File: mmdet3d/models/backbones/pointnet2_sa_ssg.py [hunk 23]
```diff
@@ -85,7 +89,7 @@ def __init__(self,
                 fp_source_channel = cur_fp_mlps[-1]
                 fp_target_channel = skip_channel_list.pop()
 
-    def forward(self, points):
+    def forward(self, points: torch.Tensor) -> Dict[str, List[torch.Tensor]]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/pointnet2_sa_ssg.py:92]: torch.Tensor -> Tensor

File: mmdet3d/models/backbones/second.py [hunk 24]
```diff
@@ -22,14 +24,15 @@ class SECOND(BaseModule):
...
-                 in_channels=128,
-                 out_channels=[128, 128, 256],
-                 layer_nums=[3, 5, 5],
-                 layer_strides=[2, 2, 2],
-                 norm_cfg=dict(type='BN', eps=1e-3, momentum=0.01),
-                 conv_cfg=dict(type='Conv2d', bias=False),
-                 init_cfg=None,
-                 pretrained=None):
+                 in_channels: int = 128,
+                 out_channels: List[int] = [128, 128, 256],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/second.py:28]: List -> Sequence

File: mmdet3d/models/backbones/second.py [hunk 25]
```diff
@@ -22,14 +24,15 @@ class SECOND(BaseModule):
...
+                 layer_nums: List[int] = [3, 5, 5],
+                 layer_strides: List[int] = [2, 2, 2],
+                 norm_cfg: Dict[str, Any] = dict(
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/second.py:31]: ConfigType

File: mmdet3d/models/backbones/second.py [hunk 26]
```diff
@@ -22,14 +24,15 @@ class SECOND(BaseModule):
...
+                     type='BN', eps=1e-3, momentum=0.01),
+                 conv_cfg: Dict[str, Any] = dict(type='Conv2d', bias=False),
+                 init_cfg: Optional[Dict[str, Any]] = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/second.py:34]: OptMultiConfig

File: mmdet3d/models/backbones/second.py [hunk 27]
```diff
@@ -75,7 +78,7 @@ def __init__(self,
         else:
             self.init_cfg = dict(type='Kaiming', layer='Conv2d')
 
-    def forward(self, x):
+    def forward(self, x: Tensor) -> Tuple[Tensor, Tensor, Tensor]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/second.py:81]: Tuple[Tensor, ...]

File: mmdet3d/structures/bbox_3d/utils.py [hunk 28]
```diff
@@ -7,6 +7,7 @@
 from torch import Tensor
 
 from mmdet3d.utils import array_converter
+from .cam_box3d import CameraInstance3DBoxes
```
Comment (Xiangxu-0103) [inline mmdet3d/structures/bbox_3d/utils.py:10]: This will cause a `circular import error`

Comment (Lum1104) [reply mmdet3d/structures/bbox_3d/utils.py:10]: Thank you for all the suggestions!!! I will fix it right now.

File: mmdet3d/models/backbones/multi_backbone.py [hunk 29]
```diff
@@ -27,16 +29,17 @@ class MultiBackbone(BaseModule):
...
-                 **kwargs):
+                 num_streams: int,
+                 backbones: Union[List, Dict],
+                 aggregation_mlp_channels: Optional[Sequence[int]] = None,
+                 conv_cfg: ConfigType = dict(type='Conv1d'),
+                 norm_cfg: ConfigType = dict(
+                     type='BN1d', eps=1e-5, momentum=0.01),
+                 act_cfg: ConfigType = dict(type='ReLU'),
+                 suffixes: Tuple = ('net0', 'net1'),
+                 init_cfg: Optional[Dict] = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/multi_backbone.py:40]: Optional[Dict] -> OptMultiConfig

File: mmdet3d/models/backbones/multi_backbone.py [hunk 30]
```diff
@@ -27,16 +29,17 @@ class MultiBackbone(BaseModule):
...
-                 pretrained=None,
-                 **kwargs):
+                 num_streams: int,
+                 backbones: Union[List, Dict],
+                 aggregation_mlp_channels: Optional[Sequence[int]] = None,
+                 conv_cfg: ConfigType = dict(type='Conv1d'),
+                 norm_cfg: ConfigType = dict(
+                     type='BN1d', eps=1e-5, momentum=0.01),
+                 act_cfg: ConfigType = dict(type='ReLU'),
+                 suffixes: Tuple = ('net0', 'net1'),
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/multi_backbone.py:39]: Tuple[str]

File: mmdet3d/models/backbones/multi_backbone.py [hunk 31]
```diff
@@ -27,16 +29,17 @@ class MultiBackbone(BaseModule):
...
-                 aggregation_mlp_channels=None,
-                 conv_cfg=dict(type='Conv1d'),
-                 norm_cfg=dict(type='BN1d', eps=1e-5, momentum=0.01),
-                 act_cfg=dict(type='ReLU'),
-                 suffixes=('net0', 'net1'),
-                 init_cfg=None,
-                 pretrained=None,
-                 **kwargs):
+                 num_streams: int,
+                 backbones: Union[List, Dict],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/multi_backbone.py:33]: Union[List[dict], dict]

File: mmdet3d/models/backbones/nostem_regnet.py [hunk 32]
```diff
@@ -59,15 +64,19 @@ class NoStemRegNet(RegNet):
         (1, 1008, 1, 1)
     """
 
-    def __init__(self, arch, init_cfg=None, **kwargs):
+    def __init__(self,
+                 arch: Dict[str, Union[int, float]],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/nostem_regnet.py:68]: Dict[str, Union[int, float]] -> dict

File: mmdet3d/models/backbones/pointnet2_sa_ssg.py [hunk 33]
```diff
@@ -31,20 +34,23 @@ class PointNet2SASSG(BasePointNet):
...
+                 num_points: Sequence[int] = (2048, 1024, 512, 256),
+                 radius: Sequence[float] = (0.2, 0.4, 0.8, 1.2),
+                 num_samples: Sequence[int] = (64, 32, 16, 16),
+                 sa_channels: Sequence[Sequence[int]] = ((64, 64, 128),
+                                                         (128, 128, 256),
+                                                         (128, 128, 256),
+                                                         (128, 128, 256)),
+                 fp_channels: Sequence[Sequence[int]] = ((256, 256), (256,
+                                                                      256)),
+                 norm_cfg: dict = dict(type='BN2d'),
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/pointnet2_sa_ssg.py:47]: ConfigType

File: mmdet3d/models/backbones/pointnet2_sa_ssg.py [hunk 34]
```diff
@@ -31,20 +34,23 @@ class PointNet2SASSG(BasePointNet):
...
+                 sa_cfg: dict = dict(
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/pointnet2_sa_ssg.py:48]: ConfigType

File: mmdet3d/models/backbones/multi_backbone.py [hunk 35]
```diff
@@ -27,16 +29,17 @@ class MultiBackbone(BaseModule):
...
+                 aggregation_mlp_channels: Optional[Sequence[int]] = None,
+                 conv_cfg: ConfigType = dict(type='Conv1d'),
+                 norm_cfg: ConfigType = dict(
+                     type='BN1d', eps=1e-5, momentum=0.01),
+                 act_cfg: ConfigType = dict(type='ReLU'),
+                 suffixes: Tuple = ('net0', 'net1'),
+                 init_cfg: Optional[Dict] = None,
+                 pretrained: Optional[str] = None,
+                 **kwargs: Any) -> None:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/multi_backbone.py:42]: For `**kwargs`, we do not need to add typehint here.

File: mmdet3d/models/backbones/nostem_regnet.py [hunk 36]
```diff
@@ -59,15 +64,19 @@ class NoStemRegNet(RegNet):
...
+                 init_cfg: OptMultiConfig = None,
+                 **kwargs: Any) -> None:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/nostem_regnet.py:70]: The same above

**Pull request merged**